### PR TITLE
Pass environment variables around as a data structure, via @sordina

### PR DESCRIPTION
**Pull request by sordina**
### Description

Environment variables are currently often accessed through IO calls to getEnvironment, and lookupEnv, etc.

This causes issues:

* Running two or more GraphQL-Engine threads in the same process with different environments is impossible
* Redundant IO calls are performed in code that could easily otherwise be pure

This is a refactoring to only require an IO call at the main entry-point and then subsequent passing of environment through explicit arguments and implicit structures and monads.

### Affected components

- [x] Server

### Solution and Design

The solution was to work bottom-up:

* The creation of a new `Data.Environment` module that exposed an IO call to create the environment, then all subsequent functions to access and lookup the environment as pure operations on `Environment`
* Replacing existing IO environment calls with pure `Data.Environment` calls.
* Resolution of errors to pass the environment as an argument to these functions
* Refactoring the argument into existing datatypes and monads in scope where appropriate

### Steps to test and verify

* Existing test-suite should ensure no regressions are made for existing behaviour.
* Explicit testing of environment accessing interfaces such as Events, Actions, etc.

### Limitations, known bugs & workarounds

The introduction of environment arguments is sometimes cumbersome. Further scrutiny of the implementation may reveal opportunities for more elegant or concise implementation through reader context, etc.

### Server checklist

* Ensure test coverage is adequate for environment-related functionality testing

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [X] No

#### Metadata

Does this PR add a new Metadata feature?
- [X] No

#### GraphQL
- [X] No new GraphQL schema is generated

#### Breaking changes

* Library usage of GraphQL-Engine may now require explicit interaction with `Data.Environment`

No UI driven braking changes have been introduced.

### CODE REVIEW

File: server/src-lib/Hasura/Eventing/EventTrigger.hs [hunk 1]
```diff
@@ -158,45 +158,56 @@ initEventEngineCtx maxT _eeCtxFetchInterval = do
...
-             link t
-
-        -- return when next batch ready; some 'processEvent' threads may be running.
-        wait eventsNextA
+          processEvent event
+            & withEventEngineCtx eeCtx
+            & flip runReaderT (logger, httpMgr)
+          -- removing an event from the _eeCtxLockedEvents after the event has
+          -- been processed
+          removeEventFromLockedEvents (eId event) leEvents
```
Comment (0x777) [inline server/src-lib/Hasura/Eventing/EventTrigger.hs:209]: looks like the `link t` part is missing?

File: server/src-lib/Hasura/Logging.hs [hunk 2]
```diff
@@ -108,6 +108,7 @@ data InternalLogTypes
  | ILTJwkRefreshLog
  | ILTTelemetry
  | ILTSchemaSyncThread
+ | ILTPollerLog
```
Comment (0x777) [inline server/src-lib/Hasura/Logging.hs:111]: Shouldn't be present, probably a bad merge somewhere up the chain.

File: server/src-lib/Hasura/Eventing/EventTrigger.hs (hunk 1 above)
Comment (sordina) [reply server/src-lib/Hasura/Eventing/EventTrigger.hs:209]: Good catch... Strange that that got removed.

Comment (sordina) [reply server/src-lib/Hasura/Eventing/EventTrigger.hs:209]: Ah actually, is there a thread to be linked?

Comment (sordina) [reply server/src-lib/Hasura/Eventing/EventTrigger.hs:209]: Seems like this action is no longer performed async, so there's nothing to link.

File: server/src-lib/Hasura/Logging.hs (hunk 2 above)
Comment (sordina) [reply server/src-lib/Hasura/Logging.hs:111]: So remove ILTPollerLog references? There's also a case-match on line 126.

File: server/src-lib/Hasura/Server/Logging.hs [hunk 3]
```diff
@@ -25,6 +25,13 @@ import qualified Data.Text                 as T
 import qualified Network.HTTP.Types        as HTTP
 import qualified Network.Wai.Extended      as Wai
 
+-- import qualified Data.ByteString.Lazy                   as BL
```
Comment (0x777) [inline server/src-lib/Hasura/Server/Logging.hs:28]: We can remove these?

Comment (sordina) [reply server/src-lib/Hasura/Server/Logging.hs:28]: Don't see why not. We can always reimport them if they're required.

File: server/src-lib/Hasura/Server/Auth.hs [hunk 4]
```diff
@@ -1,4 +1,6 @@
-{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
```
Comment (0x777) [inline server/src-lib/Hasura/Server/Auth.hs:2]: `RecordWildCards` and `GeneralizedNewtypeDeriving` is already part of cabal's `default-extensions`.

Comment (sordina) [reply server/src-lib/Hasura/Server/Auth.hs:2]: Should I remove all references to these extensions? Or just from my PR?

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 5]
```diff
@@ -120,22 +122,22 @@ runHasuraGQ
...
   (E.ExecutionCtx logger _ pgExecCtx _ _ _ _ _) <- ask
-  logQuery' logger
-  (telemTimeIO, respE) <- withElapsedTime $ liftIO $ runExceptT $ case resolvedOp of
-    E.ExOpQuery tx _genSql -> do
+  (telemTimeIO, respE) <- withElapsedTime $ runExceptT $ case resolvedOp of
+    E.ExOpQuery tx genSql _asts -> do
       -- log the generated SQL and the graphql query
-      -- L.unLogger logger $ QueryLog query genSql reqId
-      ([],) <$> runQueryTx pgExecCtx tx
+      logQueryLog logger query genSql reqId
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs:136]: @sordina Do you have any idea on what these changes are related to?

File: server/src-lib/Hasura/Eventing/EventTrigger.hs (hunk 1 above)
Comment (tirumaraiselvan) [reply server/src-lib/Hasura/Eventing/EventTrigger.hs:209]: This was a regression in an earlier master. `processEvent` needs to be async.

File: server/src-lib/Hasura/GraphQL/Execute/Query.hs [hunk 6]
```diff
@@ -196,51 +202,61 @@ convertQuerySelSet
...
-  => HTTP.Manager
+  => Env.Environment
+  -> HTTP.Manager
   -> [N.Header]
   -> QueryReusability
   -> V.ObjectSelectionSet
   -> QueryActionExecuter
-  -> m (LazyRespTx, Maybe ReusableQueryPlan, GeneratedSqlMap)
-convertQuerySelSet manager reqHdrs initialReusability selSet actionRunner = do
+  -> m (tx EncJSON, Maybe ReusableQueryPlan, GeneratedSqlMap, [R.QueryRootFldUnresolved])
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Execute/Query.hs:214]: Is this related to caching?

File: server/src-lib/Hasura/Eventing/EventTrigger.hs (hunk 1 above)
Comment (sordina) [reply server/src-lib/Hasura/Eventing/EventTrigger.hs:209]: Ah ok. I'll try patching this.

File: server/src-lib/Hasura/GraphQL/Execute/Query.hs (hunk 6 above)
Comment (sordina) [reply server/src-lib/Hasura/GraphQL/Execute/Query.hs:214]: This does appear to be related to caching. I think that maybe it was included accidentally when splitting the PR logically into project, env, and tracing/caching.

File: server/src-lib/Hasura/Eventing/EventTrigger.hs (hunk 1 above)
Comment (sordina) [reply server/src-lib/Hasura/Eventing/EventTrigger.hs:210]: I've put in an async call and link again:


```
      eventsNext <- LA.withAsync popEventsBatch $ \eventsNextA -> do
        -- process approximately in order, minding HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE:
        forM_ events $ \event -> do
          t <- processEvent event
            & withEventEngineCtx eeCtx
            & flip runReaderT (logger, httpMgr)
            & LA.async
          -- removing an event from the _eeCtxLockedEvents after the event has
          -- been processed
          removeEventFromLockedEvents (eId event) leEvents
          LA.link t
        LA.wait eventsNextA
```

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs (hunk 5 above)
Comment (sordina) [reply server/src-lib/Hasura/GraphQL/Transport/HTTP.hs:136]: This is probably a leaked bit of the tracing implementation.

File: server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs [hunk 7]
```diff
@@ -340,19 +344,24 @@ onStart serverEnv wsConn (StartMsg opId q) = catchAndIgnore $ do
...
+      -> GQLReqUnparsed
+      -> UserInfo
+      -> E.ExecOp (LazyTx QErr)
+      -> ExceptT () m ()
     runHasuraGQ timerTot telemCacheHit reqId query userInfo = \case
-      E.ExOpQuery opTx genSql ->
-        execQueryOrMut Telem.Query genSql $ runQueryTx pgExecCtx opTx
+      E.ExOpQuery opTx genSql _asts ->
+        execQueryOrMut Telem.Query genSql . fmap snd $
+          ([],) <$> runLazyTx pgExecCtx Q.ReadOnly opTx
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs:358]:
```suggestion
        execQueryOrMut Telem.Query genSql $ runQueryTx pgExecCtx opTx 
```

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 8]
```diff
@@ -120,22 +122,22 @@ runHasuraGQ
...
+      ([],) <$> runLazyTx pgExecCtx Q.ReadOnly tx
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs:137]:
```suggestion
      ([],) <$> runQueryTx pgExecCtx tx
```
